tasks/indexing/task_ipni.go

```diff
@@ -0,0 +1,454 @@
+package indexing
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	"github.com/filecoin-project/curio/lib/storiface"
+	"github.com/ipfs/go-cid"
+	logging "github.com/ipfs/go-log/v2"
+	carv2 "github.com/ipld/go-car/v2"
+	"github.com/ipld/go-car/v2/index"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	"github.com/ipni/go-libipni/ingest/schema"
+	"github.com/ipni/go-libipni/metadata"
+	"github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-state-types/abi"
+
+	"github.com/filecoin-project/curio/deps/config"
+	"github.com/filecoin-project/curio/harmony/harmonydb"
+	"github.com/filecoin-project/curio/harmony/harmonytask"
+	"github.com/filecoin-project/curio/harmony/resources"
+	"github.com/filecoin-project/curio/lib/ffi"
+	"github.com/filecoin-project/curio/lib/indexstore"
+	"github.com/filecoin-project/curio/lib/ipni/chunker"
+	"github.com/filecoin-project/curio/lib/ipni/ipniculib"
+	"github.com/filecoin-project/curio/lib/passcall"
+	"github.com/filecoin-project/curio/lib/pieceprovider"
+)
+
+var ilog = logging.Logger("ipni")
+
+type IPNITask struct {
+	db            *harmonydb.DB
+	indexStore    *indexstore.IndexStore
+	pieceProvider *pieceprovider.PieceProvider
+	sc            *ffi.SealCalls
+	cfg           *config.CurioConfig
+}
+
+func NewIPNITask(db *harmonydb.DB, sc *ffi.SealCalls, indexStore *indexstore.IndexStore, pieceProvider *pieceprovider.PieceProvider, cfg *config.CurioConfig) *IPNITask {
+
+	return &IPNITask{
+		db:            db,
+		indexStore:    indexStore,
+		pieceProvider: pieceProvider,
+		sc:            sc,
+		cfg:           cfg,
+	}
+}
+
+func (I *IPNITask) Do(taskID harmonytask.TaskID, stillOwned func() bool) (done bool, err error) {
+	ctx := context.Background()
+
+	var tasks []struct {
+		SPID   int64                   `db:"sp_id"`
+		Sector abi.SectorNumber        `db:"sector"`
+		Proof  abi.RegisteredSealProof `db:"reg_seal_proof"`
+		Offset int64                   `db:"sector_offset"`
+		CTXID  []byte                  `db:"context_id"`
+		RM     bool                    `db:"is_rm"`
+		Prov   string                  `db:"provider"`
+	}
+
+	err = I.db.Select(ctx, &tasks, `SELECT 
+											sp_id, 
+											sector, 
+											reg_seal_proof,
+											sector_offset,
+											context_id,
+											is_rm,
+											provider	
+										FROM 
+											ipni_task
+										WHERE 
+											task_id = $1;`, taskID)
```

Comment:
Since only one row is returned, it seems more efficient to use `db.QueryRow` and then `db.StructScan`

Reply:
We expect only 1 row but it can return more. Thus the condition check below this. The DB is a custom wrapper of pgx. It kind of allows to neatly query everything into an array.